dev/env.go

```diff
@@ -0,0 +1,74 @@
+package dev
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/joho/godotenv"
+	"github.com/puma/puma-dev/homedir"
+)
+
+// load paths in order of preference, first to set a variable wins
+// `.pumaenv` has highest priority, `~/.powconfig` has lowest.
+var AppEnvPaths = []string{".env", ".powrc", ".powenv", ".pumaenv"}
+var HomeEnvPaths = []string{"~/.powconfig", "~/.pumaenv"}
+
+func LoadEnv(baseEnv map[string]string, dir string) (envMap map[string]string, err error) {
+	// build a list of all the supported env files that exist on the filesystem to load
+	var envPaths []string
+
+	for _, path := range AppEnvPaths {
+		fullPath := filepath.Join(dir, path)
+		if _, err := os.Stat(fullPath); err == nil {
+			envPaths = append(envPaths, fullPath)
+		}
+	}
+
+	for _, path := range HomeEnvPaths {
+		if fullPath, err := homedir.Expand(path); err == nil {
+			if _, err := os.Stat(fullPath); err == nil {
+				envPaths = append(envPaths, fullPath)
+			}
+		}
+	}
+
+	envMap = baseEnv
+
+	// if we have any optional env paths to source, source them
+	if len(envPaths) > 0 {
+		// load the env into a map
+		appEnv, err := godotenv.Read(envPaths...)
+		if err != nil {
+			// if any path fails to load, bail out.
+			return nil, fmt.Errorf("LoadEnv failed. %s", err.Error())
```

Comment:
This breaks a few apps of mine. We have a few `.powrc` files that are full shell scripts that try to detect if you're using `rvm` or `chruby` and do the correct initialization steps for those tools.

This change, if I understand it, essentially makes it so that the `envPaths` cannot execute scripts, only define ENV vars statically. For my team, we can change our scripts so that we can workaround this, but I can't speak for other teams. My guess is that if folks really need to execute scripts to get the right environment for puma, they can use something like `direnv` now since this PR will load the correct `$SHELL`.

Regardless, this seems like a breaking change and warrants a major version bump IMO.

Reply:
Ah, an excellent point. And merits re-thinking whether this approach is right... I hadn't considered using full shell powers in these resource files. 

I believe we could do the env loading on the Go side and then do it again in shell if we have to... Might result in weird double-PATH issues if you're doing really exotic stuff, but that seems less problematic than breaking current workflows.